### DISCUSSION TITLE
max_depth filter to query COPC overviews

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#101] [General]
Chapeau, I just tried the `-max_depth` filter which was mentioned in https://github.com/r-lidar/lidRviewer/issues/23#issuecomment-2524084610 and it works pretty decent as you can see in the example below 🚀 

But I didnt find any documentation on that, is this a hidden gem by purpose?

Another question not related to lasR, at least until we can write COPC, but you might know anyway: how is decided which points go in upper octree nodes of COPCs (could be e.g. random, highest, ... within node extent)?

``` r
library(lasR)
#> Warning: package 'lasR' was built under R version 4.4.2

f <- fs::file_temp(ext = ".copc.laz")

options(timeout=300)
download.file(url = 'https://data.cyverse.org/dav-anon/iplant/home/jgillan/USGA/imagery_products/hole17_point_cloud.copc.laz',
              destfile = f,
              mode = 'wb')

start <- Sys.time()
pipeline <- reader_las() + rasterize(10, "zmax")
ans <- exec(pipeline, on = f)
end <- Sys.time()
print(end-start)
#> Time difference of 51.36065 secs

terra::plot(ans)
```

![](https://i.imgur.com/UEv0pRz.png)

``` r


start <- Sys.time()
pipeline <- reader_las(filter = "-max_depth 0") + rasterize(10, "zmax")
ans <- exec(pipeline, on = f)
end <- Sys.time()
print(end-start)
#> Time difference of 0.06015301 secs

terra::plot(ans)
```

![](https://i.imgur.com/CYgqNgO.png)

<sup>Created on 2024-12-07 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by Jean-Romain** [comment]
> But I didnt find any documentation on that, is this a hidden gem by purpose?

Again, same answer than yesterday, if I were able to work 37 hours per day on it a lot of stuff would be more powerful, better documented, more user friendly and so on. I'm not limited by ideas or skill but by time. I'm only working a few hours per week from time to time when I need a feature myself. That is the main reason. Feel free to make a PR.

By the way, as you know I'm using LASlib to read and write LAS/LAZ files. So `max_depth` is also in `LAStools` (I did the implementation).

> at least until we can write COPC

You can: `write_las(ofile = "path/to/file.copc.laz")`

> how is decided which points go in upper octree nodes of COPCs

The best way is with a Poisson sampling. Because it is computationally demanding the implementation (**mine**) is based on voxel sampling. The root is subdivided in 128 x 128 x 128 voxels (or maybe 256). The first points that occupies a voxel takes the place. If another point wants to occupy that voxel it  is rejected and we retry at next level. The next level has 128x128x 120 smaller voxels. In practice because ALS points are usually sorted in the acquisition order this approach would lead to a visually ugly results with strong artifacts. Point are consequently inserted in a random order in order to add some entropy. As a conclusion, the final answer is that the each level has random points, BUT the voxel sampling guarantee the maximum amount of detail in each level and especially it guarantee that noise points are visible at root level which would very unlikely in a true random sampling.

This is for MY implementation. Other software may opt for a different strategy. The COPC standard does not say anything about the sampling strategy. So, for your file I cannot tell how the COPC hierarchy was created.

**Reply by Jean-Romain** [reply]
@wiesehahn in latest release `reader()` supersed `reader_las()` and has an argument `copc_depth`

**Reply by wiesehahn** [reply]
oh yes I have seen that. Just out of curiosity and to be prepared, did you think about implementing this strategy to write files as well? e.g. `writer()` instead of `write_las()`?

**Reply by Jean-Romain** [reply]
no, `reader()` can automatically leverage the correct reader depending on the file format. But it will always be up to the user to choose the output format. If I add an automatic `writer()` this won't invalidate `write_las`. `reader` supersed `reader_las()` because I'm intending to support more formats transparently. You can already use binary PCD files but the support is still very partial (this is why it is not documented)

```r
pc = read_cloud("file.pcd")
ans = exec(pipeline, on = "file.pcd")
```
